core/src/main/java/com/coderanch/blackjack/Hand.java

```diff
@@ -0,0 +1,108 @@
+/*
+ * Copyright (C) 2018 Coderanch.
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+package com.coderanch.blackjack;
+
+import java.util.*;
+
+import static com.coderanch.util.require.Require.requireThat;
+import static org.hamcrest.Matchers.is;
+import static org.hamcrest.Matchers.notNullValue;
+
+/**
+ * A hand in a game of Blackjack.
+ */
+final class Hand {
+
+    /**
+     * Highest score in Blackjack.
+     */
+    private static final int MAX_LEGAL_SCORE = 21;
+
+    /**
+     * Max amount of aces
+     */
+    private static final int MAX_SCORE_COUNT = 4;
+
+
+    private List<Card> cards;
+    private List<Integer> possibleScores;
+    private int bestScore;
+
+    /**
+     * Creates a new hand.
+     *
+     * @param card  first card dealt.
+     * @param card2 second card dealt
+     * @throws NullPointerException if either {@code card} or {@code card2} is {@code null}.
+     */
+    public Hand(Card card, Card card2) {
+        this(List.of(card, card2));
+    }
+
+    private Hand(List<Card> cards) {
+        this.cards = cards;
+        this.possibleScores = calculateScores();
+        this.bestScore = calculateBestScore();
+    }
+
+    /**
+     * Gets a new hand with the extra card.
+     *
+     * @param card the card being added to the hand.
+     * @return a new hand with the extra card.
+     * @throws IllegalArgumentException if {@code card} is {@code null}.
+     */
+    public Hand withAdditionalCard(Card card) {
+        requireThat("card", card, is(notNullValue()));
+
+        var newCards = new ArrayList<Card>(this.cards);
+        newCards.add(card);
+        return new Hand(newCards);
+    }
+
+    private List<Integer> calculateScores() {
+        var startingScore = cards.stream()
+                .filter(c -> c.rank() != Card.Rank.ACE)
+                .mapToInt(Card::points)
+                .sum();
+
+        var scores = new ArrayList<Integer>(MAX_SCORE_COUNT);
+        scores.add(startingScore);
+
+        cards.stream()
+                .filter(c -> c.rank() == Card.Rank.ACE)
+                .forEach((Card c) -> {
+                    var initialSize = scores.size();
+                    for (int i = 0; i < initialSize; i++) {
+                        var score = scores.get(i);
+                        scores.set(i, score + 1);
+                        if (i + 1 == initialSize) {
+                            scores.add(score + c.points());
+                        }
+                    }
+                });
+        return scores;
+    }
+
+    private int calculateBestScore() {
```

Comment:
You can easily calculate the best score if you write a recursive function that looks like this:

```suggestion
    private static int calculateBestScore(Collection<Integer> fixedPointValues, int numberOfFreeAces) {
```
In the recursive step, convert free aces to fixed point values.

The base case is when you don't have free aces to convert to fixed point values.

If there are no legal scores, return the lowest score (having converted all aces to 1).

Reply:
I'm stumped on using fixed point values for this.  Maybe you can give me a hint.

Reply:
Sure.

```java
    public int bestScore() {
        var minimumScore = cards.stream()
            .map(Card::rank)
            .filter(not(ACE::equals))
            .mapToInt(Rank::points)
            .sum();
        
        var numberOfFreeAces = (int) cards.stream()
            .map(Card::rank)
            .filter(ACE::equals)
            .count();
        
        return calculateBestScore(minimumScore, numberOfFreeAces);
    }

    private static int calculateBestScore(int minimumScore, int numberOfFreeAces) {
        // You still need to handle the base case here.
        
        var bestScoreWithBigAce = calculateBestScore(minimumScore + 11, numberOfFreeAces - 1);
        
        // You still need to do some more work here.
    }
```